Add config flow for One wire

epenet opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds the ability to configure the onewire component via the configuration flow.

Blocked by:
- [x] ~~#40943~~

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
sensor:
   - platform: onewire
     host: localhost
     port: 4304
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14328

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

epenet committed:
Add support for config_flow
Add support for switches
Add support for binary sensors

epenet committed:
Add config flow strings

epenet committed:
Update .coveragerc

epenet committed:
black-formatting

epenet committed:
fixes for isort and flake

epenet committed:
fixes for pylint

epenet committed:
fixes for isort

epenet committed:
fixes for isort

epenet committed:
fixes for config_flow

epenet committed:
Add devices to Device Registry

Misiu commented on homeassistant/components/onewire/__init__.py:
```diff
@@ -1 +1,28 @@
+
+
+async def async_setup(hass, config):
+    """Old way of setting up OneWire integrations."""
```
comments are inconsistent. Sometimes you write OneWire and sometimes 1-Wire.

epenet replied:
Updated in bf99fce

epenet committed:
Updated comments

epenet commented on homeassistant/components/onewire/binary_sensor.py:
```diff
@@ -0,0 +1,113 @@
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Old way of setting up deCONZ platforms."""
```
Updated in bf99fce

epenet commented on homeassistant/components/onewire/const.py:
```diff
@@ -0,0 +1,50 @@
+"""Constants for the deCONZ component."""
```
Updated in bf99fce

epenet commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -107,217 +58,119 @@ def hb_info_from_type(dev_type="std"):
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Old way of setting up deCONZ platforms."""
```
Updated in bf99fce

epenet commented on homeassistant/components/onewire/switch.py:
```diff
@@ -0,0 +1,139 @@
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Old way of setting up deCONZ platforms."""
```
Updated in bf99fce

Misiu commented on homeassistant/components/onewire/onewireentity.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class OneWireEntity(Entity):
+    """Implementation of a 1-wire Entity."""
```
```suggestion
    """Implementation of a 1-Wire Entity."""
```

epenet replied:
Updated in 91674ed2f4a5d563c09230231656aeb29a074b38

epenet commented on homeassistant/components/onewire/strings.json:
```diff
@@ -0,0 +1,30 @@
+    "abort": {},
+    "error": {
+      "connection_error": "Unable to connect to OWServer.",
+      "invalid_path": "Directory not found."
```
Hi @Misiu 
I'm not sure what you mean by that.
I'm letting the user select the connection type (OWServer using host/port, OWFS by specifying a mount directory or SYSBUS)

Are you implying that I should hide SYSBUS if the DEFAULT_MOUNT_DIR (/sys/bus/w1/devices/) doesn't exist?

Misiu replied:
I missed the error message about dtoverlay. I think this will be enough.
Don't hide that option. It's better to get an error that might help than to confuse the user by hiding the option.

epenet committed:
fixes for flake8

epenet committed:
Updated comments

epenet committed:
Updated comments

epenet commented:
@bdraco do you think I need to add a "reload" or a "rescan" service?

bdraco commented:
> @bdraco do you think I need to add a "reload" or a "rescan" service?

Config entries get reload for free as long as you have implemented unloading the config entry. It's available in the UI in 0.115dev

Misiu commented:
@epenet I guess you must add `async_unload_entry` to `__init__.py`.
Take a look at https://github.com/home-assistant/core/blob/f302c6fd650fdbb7009796db97a97aa0d6eb9001/homeassistant/components/mill/__init__.py#L20

@bdraco please correct me if I'm wrong.

epenet committed:
Implement async_unload_entry

epenet commented:
Does dc89fd1 solve this correctly?

bdraco commented:
> Does [dc89fd1](https://github.com/home-assistant/core/commit/dc89fd119ff5df2316bba6c7593e110f1c2f6b93) solve this correctly?

Without digging into all the details, it looks right at first glance 👍

ctalkington commented on .coveragerc:
```diff
@@ -597,7 +597,7 @@ omit =
     homeassistant/components/ohmconnect/sensor.py
     homeassistant/components/ombi/*
-    homeassistant/components/onewire/sensor.py
+    homeassistant/components/onewire/*
```
```suggestion
    homeassistant/components/onewire/__init__.py
```

ctalkington replied:
config flow requires 100% coverage ;)

bdraco replied:
`pytest --cov=homeassistant/components/onewire/ --cov-report term-missing -- tests/components/onewire/test_*.py`

epenet replied:
I have merged in the suggested changes from @ctalkington.
Do I need to include also the __init__.py, config_flow.py and const.py?

epenet replied:
Coverage failed, so added those back in: init.py, config_flow.py and const.py

epenet replied:
@ctalkington 
I have now added config_flow tests in 91f91fe. Are you happy to mark this as resolved?

ctalkington commented on homeassistant/components/onewire/binary_sensor.py:
```diff
@@ -0,0 +1,103 @@
+"""Support for 1-Wire environment binary sensors."""
```
This reallyreally should be done in separate PR to ease review of different features.

epenet replied:
Removed in 45701c486c6a7e04cbe0d681f6b7f3ffded82976

ctalkington commented on homeassistant/components/onewire/switch.py:
```diff
@@ -0,0 +1,130 @@
+"""Support for 1-Wire switches."""
```
this should also be part of separate PR, the general approach is one platform or feature per PR

epenet replied:
Removed in 45701c486c6a7e04cbe0d681f6b7f3ffded82976

ctalkington reviewed (changes requested):
left comments on bigger items i noticed

epenet committed:
remove binary_sensor and switch implementation (will move to new PR)

epenet committed:
Update .coveragerc

Co-authored-by: Chris Talkington <chris@talkingtontech.com>

epenet committed:
Update .coveragerc

Misiu commented:
@epenet I'm testing your changes right now. I use DS18B20 connected to Pi4 (so only SysBus option).
Generally, everything works great. I've connected two sensors, Added new integration, and two new devices, and two entities were discovered 💪
There are some things I noticed/I'd like to ask about:

1. Two sensors were discovered, after a while, I unconnected one (this might happen), but both were still visible in HA. The one that was left worked fine. I think the second (unplugged) should become unavailable. This way we would know that there is a connection error. What do you think?
2. After resetting my setup (2 sensors working fine) I plugged in another one. I had to delete the integration and add it again to get the third sensor. Isn't there an option to rescan?
3. After the setup of the integration, I got a success message, but nothing showed up. I had to wait a bit to see the entities:

![1wire](https://user-images.githubusercontent.com/1741838/92155982-8ffc4a80-ee28-11ea-99ee-91aa1adcfecc.gif)

This was a bit confusing. There should be a spinner and a message saying that the integration is detecting sensors (or ideally it should do that before it shows success message)

Also instead of naming the entry `/sys/bus/w1/devices/` maybe call it `SysBus` (put the connection type as the name).

epenet committed:
Review config flow to store the configuration type

epenet commented:
Hi @Misiu 

I've implemented the tweak regarding the name of the connection (should now default to SysBus)

Regarding the disconnected device, anything visible in the logs? I don't use SysBus personally.
```
logger:
  default: info
  logs:
    homeassistant.components.onewire: debug
```

Regarding adding a new device, I'm not sure how to detect it. Should it rescan the whole bus every xxx minutes? Or is there a core method available to all platforms to reload? Maybe @bdraco has an idea.

Finally, regarding the final step of the setup (Success message), I'm not doing anything specific so I guess it's a "feature" of home assistant.

epenet committed:
Add config_flow tests

bdraco commented:
> Hi @Misiu
> 

> Regarding adding a new device, I'm not sure how to detect it. Should it rescan the whole bus every xxx minutes? Or is there a core method available to all platforms to reload? Maybe @bdraco has an idea.
> 
Since you have implemented `async_unload_entry`, in 0.115 the user can reload the config entry in the UI.   In a future PR, should you want to have them appear automatically you could scan periodically and setup a dispatcher in the setup function to call `async_add_entities` when via a dispatch when it finds one.

ctalkington reviewed (changes requested):
looks like theres some merge conflicts

epenet committed:
Fix merge conflicts, and reduce size of PR to make it easier to compare/comment/merge

epenet commented:
@ctalkington
Following the discussion around multiple platforms, I realised that I should really try to keep my PR as small as possible, and I reworked the PR to keep the changes to the minimum.
Hopefully that makes it easier to review and accept.

epenet committed:
Move CONF_NAMES to constants

epenet committed:
Fix isort

epenet committed:
Merge branch 'dev' of https://github.com/home-assistant/core into OneWire

epenet committed:
Tweak to onewire logger

epenet committed:
Tweak to onewire logger for sensor

epenet committed:
Reset _LOGGER variable

epenet committed:
Merge branch 'dev' into OneWire

ctalkington commented:
looks like we havw a few more conflicts. if you can resolve those, ill have another look this week

epenet committed:
Merge branch 'dev' of https://github.com/home-assistant/core into OneWire

epenet committed:
Normalise header

epenet committed:
Normalise header

epenet commented:
Thanks @ctalkington, I've resolved the conflicts.

epenet committed:
Merge branch 'dev' into OneWire

epenet committed:
Update to use references in config flow translations

epenet committed:
Improve test coverage

epenet committed:
fixes for isort

ctalkington commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        errors = {}
+        if user_input is not None:
+            self.onewire_config.update(user_input)
+            if CONF_TYPE_OWFS == user_input[CONF_TYPE]:
```
```suggestion
            if user_input[CONF_TYPE] in (CONF_TYPE_OWFS, CONF_TYPE_SYSBUS):
```

ctalkington commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+                return await self.async_step_mount_dir()
+            if CONF_TYPE_OWSERVER == user_input[CONF_TYPE]:
+                return await self.async_step_owserver()
+            if CONF_TYPE_SYSBUS == user_input[CONF_TYPE]:
```
```suggestion
```

ctalkington commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+            if CONF_TYPE_OWSERVER == user_input[CONF_TYPE]:
+                return await self.async_step_owserver()
+            if CONF_TYPE_SYSBUS == user_input[CONF_TYPE]:
+                return await self.async_step_mount_dir()
```
```suggestion
```

epenet replied:
I think this is no longer relevent

ctalkington commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+                _LOGGER.error(
+                    "Cannot connect to owserver on %s:%d, got: %s", owhost, owport, exc
+                )
+                errors["base"] = "connection_error"
```
try to use common reference strings:

```suggestion
                errors["base"] = "cannot_connect"
```

ctalkington commented on homeassistant/components/onewire/strings.json:
```diff
@@ -0,0 +1,30 @@
+  "config": {
+    "abort": {},
+    "error": {
+      "connection_error": "[%key:common::config_flow::error::cannot_connect%]",
```
```suggestion
      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
```

ctalkington commented on tests/components/onewire/test_init.py:
```diff
@@ -0,0 +1,33 @@
+    """Test being able to unload an entry."""
+    config_entry = await setup_onewire_integration(hass)
+
+    assert await onewire.async_unload_entry(hass, config_entry)
```
we want to use core methods to do this
```suggestion
    assert len(hass.config_entries.async_entries(DOMAIN)) == 1
    assert config_entry.state == ENTRY_STATE_LOADED

    assert await hass.config_entries.async_unload(config_entry.entry_id)
    await hass.async_block_till_done()

    assert config_entry.state == ENTRY_STATE_NOT_LOADED
    assert not hass.data.get(DOMAIN)
```

ctalkington commented on tests/components/onewire/test_init.py:
```diff
@@ -0,0 +1,33 @@
+"""Tests for 1-Wire config flow."""
+from homeassistant import config_entries
+from homeassistant.components import onewire
+from homeassistant.components.onewire.const import CONF_TYPE_SYSBUS
```
```suggestion
from homeassistant.components.onewire.const import CONF_TYPE_SYSBUS
from homeassistant.config_entries import ENTRY_STATE_LOADED, ENTRY_STATE_NOT_LOADED
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+"""Tests for 1-Wire config flow."""
+from pyownet import protocol
+
+from homeassistant import config_entries, data_entry_flow
```
```suggestion
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+    CONF_TYPE_SYSBUS,
+    DOMAIN,
+)
+from homeassistant.const import CONF_HOST, CONF_PORT, CONF_TYPE
```
```suggestion
from homeassistant.config_entries import SOURCE_USER
from homeassistant.const import CONF_HOST, CONF_PORT, CONF_TYPE
from homeassistant.data_entry_flow import RESULT_TYPE_CREATE_ENTRY, RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+async def test_config_flow_owserver(hass):
+    """Test OWServer flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+async def test_config_flow_owfs(hass):
+    """Test OWFS flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+async def test_config_flow_sysbus(hass):
+    """Test SysBus flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+        user_input={CONF_TYPE: CONF_TYPE_SYSBUS},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+            user_input={CONF_MOUNT_DIR: "/mount/directory"},
+        )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
```
```suggestion
    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+            user_input={CONF_MOUNT_DIR: "/mount/invalid_directory"},
+        )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+        user_input={CONF_TYPE: CONF_TYPE_OWFS},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+            user_input={CONF_HOST: "1.2.3.4", CONF_PORT: 1234},
+        )
+
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
        assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+        user_input={CONF_TYPE: CONF_TYPE_OWSERVER},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+            user_input={CONF_MOUNT_DIR: "/sys/bus/invalid_directory"},
+        )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
```suggestion
    assert result["type"] == RESULT_TYPE_FORM
```

ctalkington commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,158 @@
+            user_input={CONF_MOUNT_DIR: "/sys/bus/directory"},
+        )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
```
```suggestion
    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
```

ctalkington commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        """Initialize 1-Wire config flow."""
+        self.onewire_config = {}
+
+    async def async_step_user(self, user_input=None):
```
how are we handling duplicate config entry detection? any unique ids? may need to do checks against existing entries for similiar configs.

epenet replied:
Add in 9f55b69255e3ac78b38489c5c9a6f4ee9f927546

ctalkington commented on homeassistant/components/onewire/__init__.py:
```diff
@@ -1 +1,32 @@
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+    unload_ok = True
+    for component in SUPPORTED_PLATFORMS:
+        unload_ok = unload_ok and await hass.config_entries.async_forward_entry_unload(
+            config_entry, component
+        )
+    return unload_ok
```
```suggestion
    unload_ok = all(
        await asyncio.gather(
            *[
                hass.config_entries.async_forward_entry_unload(entry, component)
                for component in SUPPORTED_PLATFORMS
            ]
        )
    )

    return unload_ok
```

ctalkington commented on homeassistant/components/onewire/__init__.py:
```diff
@@ -1 +1,32 @@
 """The 1-Wire component."""
+import voluptuous as vol
```
```suggestion
import asyncio
import voluptuous as vol
```

ctalkington commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+            owhost = user_input.get(CONF_HOST)
+            owport = user_input.get(CONF_PORT)
+            try:
+                owproxy = protocol.proxy(host=owhost, port=owport)
```
does these two lines make requests or do IO? if so theyll need to move to a sync function and be run in the executor

MartinHjelmare commented:
Note that we won't allow any changes to this integration that affects the parts that don't use a 3rd party library for the device interface. So realistically this blocks the config flow feature in this PR.

epenet committed:
Update async_unload_entry

epenet committed:
Update common strings

epenet committed:
Remove OWFS from config_flow

epenet committed:
Update imports

epenet committed:
Move connect attempt to executor

epenet committed:
Update common strings

epenet committed:
Prevent duplicate config entries

epenet commented:
Thanks for the review @ctalkington
I've implemented all your recommendations into 9f55b69255e3ac78b38489c5c9a6f4ee9f927546

I have also taken @MartinHjelmare 's recommendation into account and removed the config_flow for the OWFS implementation (we can ask those users to migrate to OWServer)

For the SysBus implementation, I have kept it in the config flow, and will be pushing up a separate PR with @Misiu to use a 3rd-party library. Hopefully you can either validate this PR, or mark it as "ready pending upstream"

epenet committed:
Fix isort

epenet committed:
Fix flake8

epenet committed:
Fix tests following removal of OWFS implementation

epenet committed:
Fix flake8

epenet committed:
Merge remote-tracking branch 'remotes/upstream/dev' into OneWire

epenet commented:
@MartinHjelmare / @ctalkington 
I've merged in the latest changes from PR 40943, but I have a small question regarding the schema differences between "config_flow" and "yaml" config schema.

CONF_TYPE is new for config_flow
CONF_MOUNT_DIR doesn't need to be set for OWServer, but yaml config forces a default value
CONF_PORT doesn't need to be set for SysBus/OWFS, but yaml config forces a default value

Is my implementation of the check in ["sensor.py > get_entities"](https://github.com/home-assistant/core/pull/39321/files#diff-5cd29519dcc1a225c15b4734cacfc496R146) correct?

epenet committed:
Adjust config from yaml to config-flow, and add ability to specify sysbus directory

epenet commented:
After consideration, I've tweaked the way schema differences are updated and moved it to [sensor.py > setup_platform](https://github.com/home-assistant/core/pull/39321/files#diff-5cd29519dcc1a225c15b4734cacfc496R132 instead)

epenet committed:
Ajust unique_id for config entries

epenet committed:
Fix test_config_flow

epenet committed:
Merge branch 'dev' into OneWire

MartinHjelmare commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -124,35 +125,41 @@ def hb_info_from_type(dev_type="std"):
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Old way of setting up 1-Wire platform."""
+    adjusted_config = config
```
We should import the config to a config flow here.

epenet replied:
Happy to implement - is there a good example of this?

MartinHjelmare replied:
Here's one example:
https://github.com/home-assistant/core/blob/9ee97cb213d659aa9b6149484c0a42522decba78/homeassistant/components/nut/sensor.py#L60-L67

epenet replied:
I have it working on my machine, but that means that I have to completely drop support for OWFS, and force-migrate those to OWServer which I was trying to avoid.
I'm also not sure what to do about "CONF_NAMES" which was used to specify friendly names for devices (and all associated sensors). 

For example, having the following configuration in the YAML would generate the sensor entity_id `"sensor.my_ds18b20_temperature"` instead of `"sensor.10_111111111111_temperature"`
```
"names": {
            "10-111111111111": "My DS18B20",
        }
```

Is there a way to avoid those two breaking changes?

KevinCathcart replied:
I don't see any reason you cannot import the name mapping, and use it when naming the created devices/entities, despite not supporting setting it via the UI. I'm not aware of any hard rule that requires everything in a config entry to be ui configurable. Sure normally that is desirable, but in some compatibility cases like this, I'm not sure it is necessary. Indeed I'd be slightly surprised if some other integration had not done something similar for compatibility reasons.

As long as these set device/entity names eventually get stored in the same way as if the user had changed them via the UI, then things basically just work from the user perspective.

Of course, said part of the entry should be clearly documented as temporary transition code, to avoid breaking backwards compatibility.  Later when yaml import is removed, a config entry migration can drop the mappings from the config entry, and from that point forward, name customization can only be done via the ui/api. 

I cannot comment on OWFS, beyond mentioning that I see no reason why somebody could not create some form of `pyowfs` or `owfs-ha` library based on the existing code, and update it to use that, which brings it back into compliance with the library policy, and would allow it to be configured via config flow. If that was done first, it would eliminate this whole concern.

MartinHjelmare replied:
I think it would be ok to import also the OWFS config and just forward that as an entry to the platform, including a name config item. We don't do any connection checks for it and make a comment about this case.

We also need to handle a restart and re-import of an OWFS config to an existing OWFS config entry. Ie, if the user has changed a config item for OWFS we should update the existing corresponding config entry. Probably keep track of the mound directory for each config entry, and use that to map them.

epenet replied:
New commits f855383 / a54662affe8f27c46af02982baec42229648de68 hopefully implements this correctly.

epenet committed:
Fix invalid merge

epenet committed:
Convert yaml to config_entry

epenet committed:
Update sysbus tests

epenet committed:
Tweaks to yaml import process, and add OWFS warning

epenet committed:
Enable migration of OWFS platform config to config_entry

epenet committed:
update the existing corresponding config entry on OWFS conversion

epenet committed:
Merge branch 'dev' into OneWire

epenet commented:
Hi @MartinHjelmare,
Are the latest changes regarding OWFS being deprecated sufficient to unblock this PR?

MartinHjelmare commented:
I think we can move forward with this yes. 👍

MartinHjelmare commented on homeassistant/components/onewire/__init__.py:
```diff
@@ -1 +1,37 @@
+
+from .const import DOMAIN, SUPPORTED_PLATFORMS
+
+CONFIG_SCHEMA = vol.Schema(
```
This isn't used. We can remove it.

MartinHjelmare commented on homeassistant/components/onewire/__init__.py:
```diff
@@ -1 +1,37 @@
+
+
+async def async_setup(hass, config):
+    """Old way of setting up 1-Wire integrations."""
```
The `async_setup` function is still needed. So it's not really true that it's the old way.

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,135 @@
+            try:
+                await self.hass.async_add_executor_job(protocol.proxy, owhost, owport)
+                await self.async_set_unique_id(
+                    f"{CONF_TYPE_OWSERVER}:{owhost}:{owport}"
```
We don't allow ip address as unique id as ip address may change. The unique id must both be unique and stable.

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,135 @@
+
+    async def async_step_import(self, platform_config):
+        """Handle import configuration from YAML."""
+        _LOGGER.info("Importing old config entry %s", platform_config)
```
Please remove this log message. We don't log config values after config validation.

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        if user_input:
+            self.onewire_config.update(user_input)
+            mount_dir = user_input.get(CONF_MOUNT_DIR)
+            if os.path.isdir(mount_dir):
```
This does I/O and should be scheduled on the executor.

MartinHjelmare replied:
I don't think we want to pass `None` to the isdir check.

MartinHjelmare replied:
Bump.

epenet replied:
I/O now scheduled on the executor

CONF_MOUNT_DIR now None:
- async_step_import set it to DEFAULT_SYSBUS_MOUNT_DIR if it is missing
- DATA_SCHEMA_MOUNTDIR marks it as required

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,135 @@
+
+        return self.async_show_form(
+            step_id="owserver",
+            data_schema=vol.Schema(
```
The schema is static so we can define it as a constant at the module level.

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        errors = {}
+        if user_input:
+            self.onewire_config.update(user_input)
+            owhost = user_input.get(CONF_HOST)
```
The config items are required so we should not use `dict.get`. Use `dict[key]` when we know the key is in the dict.

MartinHjelmare commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -126,35 +129,46 @@ def hb_info_from_type(dev_type="std"):
+        # This part of the implementation does not conform to policy regarding 3rd-party libraries, and will not longer be updated.
+        # https://developers.home-assistant.io/docs/creating_platform_code_review/#5-communication-with-devicesservices
+        _LOGGER.warning(
+            "OWFS implementation does not conform to policy regarding 3rd-party libraries, and will no longer be updated. Please migrate to OWServer implementation."
```
Please break long strings around 88 characters.

MartinHjelmare commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -126,35 +129,46 @@ def hb_info_from_type(dev_type="std"):
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
```
Change this to `a coroutine function `async_setup_platform`.

MartinHjelmare commented on homeassistant/components/onewire/sensor.py:
```diff
@@ -126,35 +129,46 @@ def hb_info_from_type(dev_type="std"):
     """Set up 1-Wire platform."""
-    entities = get_entities(config)
-    add_entities(entities, True)
+    entities = get_entities(config_entry.data)
```
We probably need to schedule `get_entities` on the executor.

MartinHjelmare commented on homeassistant/components/onewire/strings.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "config": {
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
```
I'd use the already configured device reference.

MartinHjelmare commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,110 @@
+        }
+
+
+async def test_config_flow_sysbus(hass):
```
Make sure to also test when an entry is already created and the same config is tried again.

epenet committed:
Remove CONFIG_SCHEMA

epenet committed:
Move data_schema to constants

epenet committed:
Remove log message

epenet committed:
Remove duplicate warning

epenet committed:
Update already_configured to use already_configured_device constant

epenet committed:
Schedule get_entities on the executor

epenet committed:
Update duplicate entry check for OWServer

epenet commented:
OWServer doesn't expose anything that could be used for a unique id.
I've therefore removed it from the config_entry, and added a manual check of the existing entries for a matching host and port.

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+            try:
+                existing_entry = self.get_existing_owserver_entry(owhost, owport)
+                if existing_entry is not None:
+                    raise AbortFlow("already_configured")
```
We can `return self.async_abort(...)` here.

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+                if existing_entry is not None:
+                    raise AbortFlow("already_configured")
+                await self.hass.async_add_executor_job(protocol.proxy, owhost, owport)
+                return self.async_create_entry(title=owhost, data=self.onewire_config)
```
Please only wrap the line that can raise in the try... except block. Use an `els:` block if needed.

epenet committed:
Review TryCatch

epenet committed:
Schedule os.path.isdir on the executor

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -85,17 +84,18 @@ async def async_step_owserver(self, user_input=None):
                     "Cannot connect to owserver on %s:%d, got: %s", owhost, owport, exc
                 )
                 errors["base"] = "cannot_connect"
+            if len(errors) == 0:
```
```suggestion
            if not errors:
```

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+        errors = {}
+        if user_input:
+            self.onewire_config.update(user_input)
+            owhost = user_input[CONF_HOST]
```
Please separate words with underscore, `onewire_host` or just `host`.

MartinHjelmare commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,263 @@
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {}
```
```suggestion
    assert not result["errors"]
```

MartinHjelmare commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,263 @@
+    # Valid server
+    with patch(
+        "homeassistant.components.onewire.config_flow.protocol.proxy",
+    ):
```
We should patch the integration setup functions, to avoid setting up the whole integration on create entry results. See our scaffold script template test code for example.

epenet replied:
I'm working on this one...

epenet replied:
New OneWireHub class created for this.
Is it ok to update __init__.py and sensor.py as a follow-up PR?

MartinHjelmare replied:
I was only referring to this:
https://github.com/home-assistant/core/blob/93841e3e0cfe8971c322b6b30cb32b7cd4abd508/script/scaffold/templates/config_flow/tests/test_config_flow.py#L21-L25

epenet replied:
How to make things complicated! Should I keep the new OneWireHub (will be useful down the line with the binary-sensor and switch PRs) or remove it again to keep this PR smaller?

epenet replied:
Can we keep the OneWireHub class anyway or do you want me to remove it?
I plan to use it with followup PRs (binary-sensor / switch platforms)

MartinHjelmare replied:
We can keep it.

epenet committed:
rename owhost/owport

epenet committed:
Update checks for empty

epenet committed:
Fix incorrect patch

epenet committed:
Move config_flow validation methods to new OneWireHub

epenet committed:
Fix typo and pre-commit

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,196 @@
+            try:
+                info = await validate_input_owserver(self.hass, user_input)
+
+                return self.async_create_entry(
```
Please move this to an `else:` block. We only want the line that can raise in the try... except block.

epenet replied:
Updated in d29384e

epenet committed:
Cleanup try/else

epenet committed:
patch async_setup/async_setup_entry

epenet committed:
Merge remote-tracking branch 'remotes/upstream/dev' into OneWire

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,196 @@
+            return self.async_create_entry(
+                title=platform_config[CONF_MOUNT_DIR], data=platform_config
+            )
+
```
Maybe raise `InvalidOneWireType` here as a fourth case to keep the return type consistent. There shouldn't be any implicit exit points of a step method.

MartinHjelmare replied:
Alternatively use an else block for the last case that will be valid.

epenet replied:
Updated with `else` to keep in line with other areas

epenet committed:
cleanup implicit exit point

MartinHjelmare commented on tests/components/onewire/test_config_flow.py:
```diff
@@ -0,0 +1,340 @@
+
+    # Invalid server
+    with patch(
+        "homeassistant.components.onewire.config_flow.OneWireHub.can_connect",
```
Don't patch our code, patch the library instead.

epenet replied:
Updated - it was part of the confusion around "OneWireHub"

epenet committed:
Fix invalid patch

MartinHjelmare reviewed (approved):
Great!

epenet commented:
Thanks Martin!

epenet commented:
Aaargh - the latest else tweak broke pylint
I'll amend to use your error suggestion

MartinHjelmare commented on homeassistant/components/onewire/config_flow.py:
```diff
@@ -0,0 +1,196 @@
+            return await self.async_step_mount_dir(platform_config)
+
+        # OWFS
+        else:  # pragma: no cover
```
Just remove this and outdent below.

epenet replied:
But then I miss the 100% coverage on config-flow - I've added InvalidOneWireType

MartinHjelmare replied:
Can't we add the coverage disable anywhere in the code above where we want it to stop?

epenet replied:
Updated again in commit 7973954
`# pragma: no cover` only seems to work on a "branch" (if/else) but I couldn't find to ignore all subsequent lines.

Instead I've just moved OWFS above SysBus and I think that did the trick

MartinHjelmare replied:
Good!

epenet committed:
cleanup implicit exit point

epenet committed:
cleanup implicit exit point

MartinHjelmare commented:
@ctalkington ok?

epenet commented:
Thanks @ctalkington , thanks @MartinHjelmare
